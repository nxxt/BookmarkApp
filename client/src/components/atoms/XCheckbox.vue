<template>
  <v-tooltip bottom>
    <template v-slot:activator="{ on, attrs }">
      <v-simple-checkbox
        class="x-checkbox"
        :value="isChecked"
        v-bind="attrs"
        :ripple="false"
        @input="onInput"
        v-on="on"
      />
    </template>
    <span>to md</span>
  </v-tooltip>
</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator'

@Component
export default class XCheckbox extends Vue {
  isChecked: boolean = false

  onInput() {
    this.isChecked = !this.isChecked
    this.$emit('input', this.isChecked)
  }

  setCheck(val: boolean) {
    this.isChecked = val
  }
}
</script>
