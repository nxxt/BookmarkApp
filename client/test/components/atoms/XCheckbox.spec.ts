import Vuetify from 'vuetify'
import { mount, createLocalVue } from '@vue/test-utils'
import XCheckbox from '@/components/atoms/XCheckbox.vue'

const localVue = createLocalVue()

describe('XCheckbox.vue', () => {
  // vuetify の準備
  let vuetify: typeof Vuetify
  beforeEach(() => {
    vuetify = new Vuetify()
  })

  const factory = (options?: any) => {
    return mount(XCheckbox, {
      localVue,
      vuetify,
      ...options
    })
  }

  it('input event', () => {
    const wrapper = factory()
    const event = jest.fn()
    const elm = wrapper.find('.x-checkbox.v-simple-checkbox')

    wrapper.vm.$on('input', event) // wrapperでイベント検知設定

    // 初期状態の検証
    expect(event).toBeDefined() // イベントがundefinedでないこと
    expect(event).toHaveBeenCalledTimes(0) // eventが発生していないこと
    expect(wrapper.vm.$data.isChecked).toBe(false)

    // [1] Simulate a click on the elm
    elm.trigger('click')

    // イベント後の検証
    expect(event).toHaveBeenCalledTimes(1) // イベントが親へ伝搬したこと
    const a = wrapper.emitted('input')
    expect(a).toBeDefined()
    if (a) expect(a[0][0]).toEqual(true) // イベントのペイロードが正しいこと

    // [2] Simulate a click on the elm
    elm.trigger('click')

    // イベント後の検証
    expect(event).toHaveBeenCalledTimes(2) // イベントが親へ伝搬したこと
    if (a) expect(a[1][0]).toEqual(false) // イベントのペイロードが正しいこと
  })

  describe('method', () => {
    it('toggle', () => {
      const wrapper = factory()
      const vm: any = wrapper.vm

      // 初期状態の検証
      expect(wrapper.vm.$data.isChecked).toBe(false)

      // [1]
      vm.setCheck(true)

      // 検証
      expect(wrapper.vm.$data.isChecked).toBe(true)

      // [2]
      vm.setCheck(false)

      // 検証
      expect(wrapper.vm.$data.isChecked).toBe(false)
    })
  })
  describe('the entire template', () => {
    it('snapshot', () => {
      const wrapper = factory()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})
