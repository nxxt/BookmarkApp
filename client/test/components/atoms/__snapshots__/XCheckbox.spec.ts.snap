// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XCheckbox.vue the entire template snapshot 1`] = `"<span class=\\"v-tooltip v-tooltip--bottom\\"><!----><div role=\\"button\\" aria-haspopup=\\"true\\" aria-expanded=\\"false\\" class=\\"x-checkbox v-simple-checkbox\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-checkbox-blank-outline theme--light\\"></i></div></span>"`;
